fl4health/utils/dataset_converter.py

```diff
@@ -131,15 +151,20 @@ def _cat_input_label(self, data: torch.Tensor, target: torch.Tensor) -> Tuple[to
         """The data converter function used for conditional autoencoders.
         This converter is used when we want to condition each data sample on its label.
         """
-        num_conditions = self.get_condition_vector_size()
-        # Create the condition vector by getting the one_hot encoded target.
-        one_hot_target = torch.nn.functional.one_hot(target, num_classes=num_conditions)
+        if self.do_one_hot_encoding:  # If condition needs to be one-hot encoded.
+            # Create the condition vector by getting the one_hot encoded target.
+            target = torch.nn.functional.one_hot(target, num_classes=self.get_condition_vector_size())
         # Concatenate the data and target (target is the condition)
         # We can flatten the data since self.data_shape is already saved.
-        # Target should be the original data.
-        return torch.cat([data.view(-1), one_hot_target]), data
+        return torch.cat([data.view(-1), target]), data
 
-    def unpack_input_condition(self, packed_data: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    def get_unpacking_function(self, packed_input: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```

Comment:
So I think this is still going to mean that the VAEs hold a copy of the dataset internally, because this has a "self" input. I think you're very close here, but you actually need to return a `Callable` here. I think something like this would work without needing to change much elsewhere in terms of functionality:
```
from functools import partial
def get_unpacking_function(self) -> Callable[[torch.Tensor], Tuple[torch.Tensor, torch.Tensor]]:
    condition_vector_size = self.get_condition_vector_size()
    data_shape = self.data_shape
    return partial(AutoEncoderDatasetConverter.unpack_input_condition, condition_vector_size, data_shape)

@staticmethod
def unpack_input_condition(
    cond_vec_size: int,
    data_shape: torch.Size,
    packed_data: torch.Tensor,
) -> Tuple[torch.Tensor, torch.Tensor]:
    """Unpacks model inputs (data and condition) from a tensor used in the training loop
    regardless of the converter function used to pack them. Unpacking relies on the size of the condition vector,
    and the original data shape which is saved before the packing process.

    Args:
        packed_data (torch.Tensor): Data tensor used in the training loop as the input to the model.

    Returns:
        Tuple[torch.Tensor, torch.Tensor]: Data in its original shape, and the condition vector
        to be fed into the model.
    """
    # We assume data is "batch first".
    x = packed_data[:, : -1 * cond_vec_size]  # Exclude the condition from input
    condition_matrix = packed_data[:, -1 * cond_vec_size :]  # Save the conditions
    # Data shape is restored in the unpacking process.
    assert data_shape is not None
    return x.view(-1, *data_shape), condition_matrix
```

Then you would do ```unpacking_function = self.autoencoder_converter.get_unpacking_function()``` to set the function.

Reply:
Just for posterity, we talked about this a bit together. My suggested rewrite would require the order change in `setup_client` that we were trying to avoid. However, we're going to override setup client for the example and set the data shape property after everything is setup. That way the order doesn't change, but we can do this function formation.